### ISSUE TITLE
Jenkins CI for arm64

### ISSUE THREAD

**Issue by michael2012z** [moby/moby#39479]
**Description**

Hello, I am an engineer working for ARM. 
I found there is not Jenkins CI run on arm64 platform, but in Jenkins inventory 2 machines are available: arm64v8-ubuntu-08 and arm64v8-ubuntu-09. 
Is there anything I can do to setup the CI.

**Comment by michael2012z** [comment]
Ping @thaJeztah and @thaJeztah , any comments? 
I am willing to contribute to CI on arm64 of Moby. Appreciate if anybody can share some guideline or suggestion to start with.

**Comment by thaJeztah** [comment]
Hi @michael2012z!

The arm(64) machines/CI were disabled at some point on pull requests, because the time to run CI (which is already very long) was taking too long to run on the ARM machines (I think it was twice as long or even longer, so 4 hours or up).

Work is in progress to improve the state of CI on this repository to reduce flakiness, reduce the duration of CI (through optimisation of the CI pipeline, triage tests, and run tests more specifically instead of "all tests" for every PR); also see https://github.com/moby/moby/projects/8

That said, I think we should look into running ARM64 again (faster / more powerful machines are likely available now).

Let me copy @dave-tucker and @psftw here, who are currently in the process of making changes to CI; also feel free to give me a "ping" on the community slack (if you're on there)

**Comment by michael2012z** [comment]
Hi, @thaJeztah 

Glad to receive your reply. Thanks for sharing the latest progress of CI.

I agree we can improve CI on arm64 in two aspects:

- Better machines. We may be able to provide additional arm64 servers for CI (via openlib). May I know the configuration of the 2 arm64 machines and the organization who provided them?

- As you mentioned, tailoring tests in CI. Testing the minimum set for every PR is a good idea to make CI more efficient. I can investigate this and discuss with community.

Just now I joined in moby-project channel.

**Comment by thaJeztah** [comment]
Sorry for the delay (was held up with other tasks); I'm not familiar myself with the configuration of those machines, but others may have more information.

We can start discussing on Slack, and if needed setup a call in a timezone that works for everyone (I'm in CET, but others are in the USA, so we'll have to find a proper time).

Setting up CI for ARM may depend on having some other changes in first (I think there may be a migration to a new Jenkins cluster), and we'll likely have to discuss who/how those machines are managed (who can access the machines, in case of issues, etc.)

**Comment by michael2012z** [comment]
That's OK, I am in yet another timezone, APAC. :) So my reply doesn't always come fast either.

Let's move more discussion to Slack, try to figure out a plan to move forward.

**Comment by thaJeztah** [comment]
Copying @michael2012z's comment from the Slack conversation;

> I think the majority of CI time was spent in integration test. I measured the time on a arm64 machine (Boardcom) in our lab. Integration test took more than 3 hours (11563.57 seconds exactly). It is still too long. The machine didn't improve much.
> And I examined the time used in test cases. See the sorted table in attached excel. I hoped to find some extremely slow test cases that we may consider to exclude on arm. But I failed. The test cases on arm64 were generally slower than that on x86. In integration test "docker in docker" is used, right? I think this is part of the reason of slow CI on arm, as "docker in docker" doesn't perform well, this is known.

> Is it possible to run a minimum set of test in arm64 CI (maybe unit tests only, or adding a small collection of integration tests)?

I also copied the test-durations in a gist; https://gist.github.com/thaJeztah/2da837e096fac3ede6f7cca110d5ed48

[moby-integration-test-time-summary.xlsx](https://github.com/moby/moby/files/3399361/moby-integration-test-time-summary.xlsx)

**Comment by michael2012z** [comment]
Based on current findings in integration tests on arm, I think tailoring test set is a more feasible way to bring back CI on arm.
I have an idea to make a minimum integration test case set:

Currently we test following folders in integration test:
./integration-cli/
./integration/network/
./integration/network/ipvlan/
./integration/network/macvlan/
./integration/image/
./integration/container/
./integration/config/
./integration/volume/
./integration/build/
./integration/plugin/
./integration/plugin/volumes/
./integration/plugin/volumes/cmd/
./integration/plugin/volumes/cmd/dummy/
./integration/plugin/graphdriver/
./integration/plugin/authz/
./integration/plugin/logging/
./integration/plugin/logging/cmd/
./integration/plugin/logging/cmd/close_on_start/
./integration/plugin/logging/cmd/dummy/
./integration/plugin/common/
./integration/system/
./integration/secret/
./integration/session/
./integration/service/
On arm we can select some of them that are most critical and most probably architecture relevant to test. For example: 
./integration-cli/
./integration/system/
./integration/image/
./integration/container/
./integration/plugin/common/
./integration/system/
(I am not sure whether this is a good selection or not, open for discussion.)

Specifically we can modify code like this to implement the minimal test set:
1. Add variable `integration_api_dirs_mini` in hack/make/.integration-test-helpers to contain the small collection of test dirs (mentioned above).
2. Add a function `run_test_integration_suites_mini()` in hack/make/.integration-test-helpers to test `integration_api_dirs_mini` folders, like https://github.com/moby/moby/blob/master/hack/make/.integration-test-helpers#L31 
3. Add file hack/make/test-integration-mini to call `run_test_integration_suites_mini`. hack/make/test-integration-mini would be quite similar to hack/make/test-integration.
4. In hack/ci/arm, replace `test-integration` with `test-integration-mini`.

Ping @thaJeztah and @psftw , could you give some comments?
I can make a PR to discuss the detail.

**Comment by michael2012z** [comment]
Update about the CI duration on ARM:

I compared the duration of test case TestBuildEnvUsage2 (./integration-cli/docker_cli_build_test.go) on different machines. It is one of the top durable cases in integration test.

This case builds an image, the Dockerfile contains several dozens of commands. I think making an image brings workload in I/O. The time used to finish the case distinguished a log between machines with SSD storage and that with HDD.
On a X86 host with SSD, the test case took 30s to finish.
On a ARM64 host with SSD, it took 43s. Slower than X86 but not too bad.
On a ARM64 host with HDD, it took 192s. Almost 5 times slower than that on ARM64 host with SSD.

The overall time used for integration took 4 hours on a HDD ARM64 host. But on a ARM64 host with SSD only 1.5 hours were needed.

In Moby CI, the 2 ARM64 machines should be equipped with SSD (https://www.packet.com/cloud/servers/c1-large-arm). Maybe we can test on them and bring back CI of ARM64 as the first step.

**Comment by michael2012z** [comment]
Hi, 

Regarding the minimal test set for ARM, I created a PR for discussion: https://github.com/moby/moby/pull/39595

The PR is different with the proposal I made days ago (https://github.com/moby/moby/issues/39479#issuecomment-512188834). 

I simplified it by skipping legacy integration test cases:

- Introduce a new environment variable SKIP_LEGACY_INTEGRATION_TEST, default value is "no". 
- If it's set to "yes" or anything rather than "no", legacy integration suites will be skipped.
- In arm CI, SKIP_LEGACY_INTEGRATION_TEST is set to "yes".
- CI for other architecture's are not impacted.

**Comment by michael2012z** [comment]
Update about the time spent in integration test.

I compared the time used in tailored integration test (mentioned above) and that with legacy cases.
I run the tests on an arm64 machine with SSD. (The arm64 hosts in project CI are also equipped with SSD.)
- When legacy integration cases were included, the test took 155 minutes, covered about 1700 cases.
- When legacy integration cases were NOT included, the test took 17 minutes, covered about 300 cases.

So, if we tailor the test on arm64 by skipping legacy integration test, we can save about 2.5 hours in CI. Considering the machine I used could be a bit powerful than the machines in project CI, more time should be saved actually.

Is community interested in this? Appreciate if anybody would like to comment.
